File: erikdyrelius-python3/day05.py
from aocbase import readInput

inp = readInput()
ex = """0
3
0
1
-3"""

def puzzle1(s):
    insts = [int(x) for x in s.splitlines()]
    pc = 0
    steps = 0
    while 0 <= pc < len(insts):
        inst = insts[pc]
        insts[pc] = inst + 1
        pc = inst + pc
        steps = steps + 1
    return steps

def puzzle2(s):
    insts = [int(x) for x in s.splitlines()]
    pc = 0
    steps = 0
    while 0 <= pc < len(insts):
        inst = insts[pc]
        npc = inst + pc
        if inst >= 3:
            insts[pc] = inst - 1
        else:
            insts[pc] = inst + 1
        pc = npc
        steps = steps + 1
    return steps

print("Puzzle 1, example:  ", puzzle1(ex))
print("Puzzle 1, solution: ", puzzle1(inp))

print("Puzzle 2, example:  ", puzzle2(ex))
print("Puzzle 2, solution: ", puzzle2(inp))

File: erikdyrelius-nim/day05.nim
import aocbase
import unittest
import strutils
import sets
import algorithm

proc puzzle1(s: string): int =
    var v: seq[int] = @[]
    for line in s.splitLines:
        v.add(line.parseInt)
    var pos = 0
    while pos >= 0 and pos < v.len:
        let newPos = pos + v[pos]
        v[pos].inc
        result.inc
        pos = newPos

proc puzzle2(s: string): int =
    var v: seq[int] = @[]
    for line in s.splitLines:
        v.add(line.parseInt)
    var pos = 0
    while pos >= 0 and pos < v.len:
        let newPos = pos + v[pos]
        if v[pos] >= 3:
            v[pos].dec
        else:
            v[pos].inc
        result.inc
        pos = newPos

let inpData = getInpData()
let ex1 = """0
3
0
1
-3"""

suite "Puzzle 1":
    test "Example 1":
        check(puzzle1(ex1) == 5)
    echo "Result: ", puzzle1(inpData)

suite "Puzzle 2":
    test "Example 1":
        check(puzzle2(ex1) == 10)
    echo "Result: ", puzzle2(inpData)